File: java-devildyw/src/main/java/com/dyw/leetcode/TOP面试题/L11盛最多水的容器.java
package com.dyw.leetcode.TOP面试题;

/**
 * @author Devil
 * @since 2022-11-29-11:42
 */
@SuppressWarnings("all")
public class L11盛最多水的容器 {

    public static void main(String[] args) {
        System.out.println(new L11盛最多水的容器().maxArea(new int[]{1, 8, 6, 2, 5, 4, 8, 3, 7}));
    }

    /**
     * 暴力解法 两层for
     * @param height
     * @return
     */
    public int maxArea01(int[] height) {
        int max = 0;
        int n = height.length;
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                max = Math.max((j-i)*Math.min(height[i],height[j]),max);
            }
        }
        return max;
    }

    /**
     * 双指针 从两边往中间寻找
     *
     * @param height
     * @return
     */
    public int maxArea(int[] height) {
        int i = 0, j = height.length - 1, res = 0;

        /**
         * 在每个状态下，无论长板或短板向中间收窄一格，都会导致水槽 底边宽度 -1 变短：
         *
         * 若向内 移动短板（i 向右移动 i 变大） ，水槽的短板 min(h[i], h[j])min(h[i],h[j]) 可能变大，因此下个水槽的面积 可能增大 。
         * 若向内 移动长板（j 向左移动 j减小） ，水槽的短板 min(h[i], h[j])min(h[i],h[j]) 不变或变小，因此下个水槽的面积 一定变小 。
         *
         */
        while (i < j) {
            res = height[i] < height[j] ? Math.max(res, (j - i) * height[i++]) : Math.max(res, (j - i) * height[j--]);
        }
        return res;
    }
}

File: golang-devildyw/middle/L11盛最多水的容器.go
package main

func main() {
	maxArea([]int{1, 2, 4, 3})
}

// 每次移动较短的那条边，因为：
//   - 宽度会减小
//   - 只有移动较短边才有可能找到更高的边，从而增加面积
//   - 移动较长边只会让容器更小（因为高度由短边决定）
func maxArea(height []int) int {
	result := 0
	l, r := 0, len(height)-1

	for l < r {
		temp := min(height[l], height[r])
		result = max(result, temp*(r-l))
		// 移动较小的那一边（贪心策略）
		if height[l] < height[r] {
			l++
		} else {
			r--
		}
	}
	return result
}
